src/main/java/kafka/Consumer.java

```diff
@@ -3,47 +3,44 @@
 import configuration.Config;
 import java.time.Duration;
 import java.util.Date;
-import java.util.stream.StreamSupport;
 import monitoring.Monitor;
-import org.apache.kafka.clients.consumer.ConsumerRecord;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.scheduler.Schedulers;
+import reactor.kafka.receiver.KafkaReceiver;
 import target.ITarget;
 
 public class Consumer {
 
-    private final ReactiveKafkaClient<String, String> kafkaConsumer;
+    private final KafkaReceiver<String, String> kafkaReceiver;
     private final ITarget target;
 
-    public Consumer(ReactiveKafkaClient<String, String> kafkaConsumer, ITarget target) {
-        this.kafkaConsumer = kafkaConsumer;
+    public Consumer(KafkaReceiver<String, String> kafkaReceiver, ITarget target) {
+        this.kafkaReceiver = kafkaReceiver;
         this.target = target;
     }
 
     public Flux<?> stream() {
-        return kafkaConsumer
-            .doOnNext(records -> Monitor.batchProcessStarted(records.count()))
+        return kafkaReceiver
+            .receiveBatch()
+            .doOnNext(records -> records.count().doOnNext(Monitor::batchProcessStarted))
             .concatMap(records -> {
                 var batchStartTimestamp = new Date().getTime();
-                var hasNullKey = StreamSupport
-                    .stream(records.spliterator(), false)
-                    .anyMatch(record -> record.key() == null);
-
-                return Flux
-                    .fromIterable(records)
-                    .groupBy(record -> hasNullKey ? record.partition() : record.key())
+                return records
+                    .groupBy(x -> x.key() == null ? x.partition() : x.key())
                     .delayElements(Duration.ofMillis(Config.PROCESSING_DELAY))
                     .publishOn(Schedulers.parallel())
-                    .flatMap(partition -> partition.concatMap(record -> Mono.fromFuture(target.call(record))))
+                    .flatMap(partition ->
+                        partition.concatMap(record -> Mono.fromFuture(target.call(record)).map(__ -> record))
+                    )
                     .collectList()
-                    .map(__ -> batchStartTimestamp);
-            })
-            .doOnNext(Monitor::batchProcessCompleted)
-            .map(__ -> {
-                kafkaConsumer.commit();
-                return 0;
+                    .doOnNext(__ -> Monitor.batchProcessCompleted(batchStartTimestamp));
             })
-            .doOnNext(__ -> kafkaConsumer.poll());
+            .flatMap(records ->
+                Mono.fromRunnable(() -> {
+                    var lastOffset = records.get(records.size() - 1).receiverOffset();
+                    lastOffset.commit().block();
```

Comment:
Isn't there something out of the box that can do this for us?

Reply:
Not that I found